### PR TITLE
Raise MSRV to Rust 1.80

### PR DESCRIPTION
**Pull request by ErichDonGubler**
- [ ] Depends on Firefox updating to Rust 1.80 or later.
- ~~Depends on #6147.~~
- ~~Depends on #6180.~~
- Part of #6379.

### CODE REVIEW

**Comment by ErichDonGubler** [comment]
Heeey, we're finally running into the same annoying errors [here](https://github.com/gfx-rs/wgpu/actions/runs/10351833426/job/28651170468?pr=6105#step:11:404) that I wasn't sure what to do with locally when I saw them:

```
error: trait `ContextId` is never used
  --> wgpu/src/context.rs:25:11
   |
25 | pub trait ContextId: Into<ObjectId> + From<ObjectId> + Debug + 'static {}
   |           ^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dead_code)]`

error: methods `device_downlevel_properties`, `device_mark_lost`, `render_bundle_encoder_multi_draw_indirect`, `render_bundle_encoder_multi_draw_indexed_indirect`, `render_bundle_encoder_multi_draw_indirect_count`, and `render_bundle_encoder_multi_draw_indexed_indirect_count` are never used
   --> wgpu/src/context.rs:190:8
    |
34  | pub trait Context: Debug + WasmNotSendSync + Sized {
    |           ------- methods in this trait
...
190 |     fn device_downlevel_properties(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
290 |     fn device_mark_lost(
    |        ^^^^^^^^^^^^^^^^
...
793 |     fn render_bundle_encoder_multi_draw_indirect(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
802 |     fn render_bundle_encoder_multi_draw_indexed_indirect(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
812 |     fn render_bundle_encoder_multi_draw_indirect_count(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
825 |     fn render_bundle_encoder_multi_draw_indexed_indirect_count(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: methods `device_downlevel_properties`, `device_mark_lost`, `render_bundle_encoder_multi_draw_indirect`, `render_bundle_encoder_multi_draw_indexed_indirect`, `render_bundle_encoder_multi_draw_indirect_count`, and `render_bundle_encoder_multi_draw_indexed_indirect_count` are never used
    --> wgpu/src/context.rs:1249:8
     |
1171 | pub(crate) trait DynContext: Debug + WasmNotSendSync {
     |                  ---------- methods in this trait
...
1249 |     fn device_downlevel_properties(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1349 |     fn device_mark_lost(&self, device: &ObjectId, device_data: &crate::Data, message: &str);
     |        ^^^^^^^^^^^^^^^^
...
1790 |     fn render_bundle_encoder_multi_draw_indirect(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1799 |     fn render_bundle_encoder_multi_draw_indexed_indirect(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1809 |     fn render_bundle_encoder_multi_draw_indirect_count(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1822 |     fn render_bundle_encoder_multi_draw_indexed_indirect_count(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    Checking naga-fuzz v0.0.0 (/Users/runner/work/wgpu/wgpu/naga/fuzz)
    Checking d3d12 v22.0.0 (/Users/runner/work/wgpu/wgpu/d3d12)
error: could not compile `wgpu` (lib) due to 3 previous errors
```

☝🏻 @cwfitzgerald, @Wumpf: Not sure how to best proceed here, but I haven't given it many brain cycles, either. Will dedicate more in the absence of feedback in the next day or so.

**Comment by ErichDonGubler** [comment]
Leaving this as a draft until we actually have the option to upgrade to Rust 1.80.

We might be able to hoist the upgrade to `cfg_aliases` out if https://github.com/katharostech/cfg_aliases/pull/10/ merges and is released.

**Comment by ErichDonGubler** [comment]
Updated OP with a link to Firefox tracking of Rust 1.80 in `mozilla-central`.

**Comment by kornelski** [comment]
`size_of` could be made backwards-compatible by importing it `use std::mem::size_of;`. Rust doesn't warn about shadowing the prelude (yet).

**Comment by ErichDonGubler** [comment]
> `size_of` could be made backwards-compatible by importing it `use std::mem::size_of;`. Rust doesn't warn about shadowing the prelude (yet).

Relevant `clippy` lint proposal by an individual who we know in this community (👋🏻 hi @kpreid!): https://github.com/rust-lang/rust-clippy/issues/8439

File: naga/src/back/msl/mod.rs [hunk 1]
```diff
@@ -668,6 +668,6 @@ pub fn write_string(
 
 #[test]
 fn test_error_size() {
-    use std::mem::size_of;
+    use size_of;
```
Comment (kornelski) [inline naga/src/back/msl/mod.rs]: Looks like find-replace error

Comment (ErichDonGubler) [reply naga/src/back/msl/mod.rs]: Fixed. 😅 Thanks!

**Comment by cwfitzgerald** [comment]
We can bump the repo msrv without the core msrv, so this doesn't need to block on ff

**Comment by ErichDonGubler** [comment]
@cwfitzgerald: This PR contains strictly backwards-incompatible changes, including in `wgpu-core` and `wgpu-hal` here. That's why I've kept this a draft and split out backwards-compatible changes repeatedly, trying to make CI green. I suppose we could reasonably split out the `wgpu`- and `examples`-specific changes for a separately mergeable PR! I don't think I'll be allocating the bandwidth for that for at least a week, but I'd be happy to review somebody who might do that before I did.

**Comment by ErichDonGubler** [comment]
Not sure why GL tests are failing. I suspect this PR is at fault, but I'm not sure yet.

**Review by cwfitzgerald — APPROVED** [review]
We seem to have UB in the GL backend that rust is now catching.

**Pull request closed without merging**